internal/ui/desk.go

```diff
@@ -96,86 +63,130 @@ func (l *deskLayout) MinSize(objects []fyne.CanvasObject) fyne.Size {
+	desk := l.app.NewWindow(fmt.Sprintf("%s%s", RootWindowName, outputName))
 	desk.SetPadded(false)
-	desk.FullScreen()
+	desk.SetFullScreen(true)
 
 	return desk
 }
 
 func (l *deskLayout) updateBackgrounds(path string) {
-	for _, background := range l.backgrounds {
-		background.updateBackgroundPath(path)
-		canvas.Refresh(background)
+	for bg := range l.backgroundScreenMap {
+		bg.updateBackgroundPath(path)
+		canvas.Refresh(bg)
 	}
 }
 
-func (l *deskLayout) Root() fyne.Window {
-	if l.win != nil {
-		return l.win
-	}
-
-	l.win = l.newDesktopWindow()
-	if l.win.Canvas().Scale() != l.screens.Scale() {
-		l.win.Canvas().SetScale(l.screens.Scale())
+func (l *deskLayout) createRoot(screen *desktop.Screen) {
+	win := l.newDesktopWindow(screen.Name)
+	l.roots = append(l.roots, win)
+	bg := newBackground()
+	l.backgroundScreenMap[bg] = screen
+	if screen == l.screens.Primary() {
+		l.primaryWin = win
+		l.bar = newBar(l)
+		l.widgets = newWidgetPanel(l)
+		l.mouse = newMouse()
+		l.mouse.Hide() // temporarily we do not draw mouse (using X default)
+		win.SetOnClosed(func() {
+			if !l.refreshing {
+				l.controlWin.Close()
+			}
+		})
+		win.SetContent(fyne.NewContainerWithLayout(l, bg, l.bar, l.widgets, l.mouse))
+		l.mouse.Hide()
+	} else {
+		win.SetContent(fyne.NewContainerWithLayout(l, bg))
 	}
+	win.Show()
+}
 
-	l.backgrounds = append(l.backgrounds, newBackground())
-	l.screenBackgroundMap[l.screens.Screens()[0]] = l.backgrounds[0]
-	l.bar = newBar(l)
-	l.widgets = newWidgetPanel(l)
-	l.mouse = newMouse()
-	l.container = fyne.NewContainerWithLayout(l, l.backgrounds[0])
-	if l.screens.Screens() != nil && len(l.screens.Screens()) > 1 {
-		for i := 1; i < len(l.screens.Screens()); i++ {
-			l.backgrounds = append(l.backgrounds, newBackground())
-			l.screenBackgroundMap[l.screens.Screens()[i]] = l.backgrounds[i]
-			l.container.AddObject(l.backgrounds[i])
+func (l *deskLayout) setupRoots() {
+	if len(l.roots) > 0 {
+		if len(l.screens.Screens()) >= len(l.roots) {
+			diff := len(l.screens.Screens()) - len(l.roots)
+			count := len(l.screens.Screens()) - diff - 1
+			for i := 0; i < diff; i++ {
+				l.createRoot(l.screens.Screens()[count])
+				count++
+			}
+		} else {
+			diff := len(l.roots) - len(l.screens.Screens())
+			count := len(l.roots) - diff - 1
+			for i := 0; i < diff; i++ {
+				root := l.roots[count]
+				root.SetOnClosed(nil)
+				bg := root.Content().(*fyne.Container).Objects[0].(*background)
+				delete(l.backgroundScreenMap, bg)
+				root.Close()
+			}
+			l.roots = l.roots[:len(l.screens.Screens())]
 		}
+		for i, root := range l.roots {
+			screen := l.screens.Screens()[i]
+			root.Hide()
+			root.SetOnClosed(nil)
+			root.SetTitle(fmt.Sprintf("%s%s", RootWindowName, screen.Name))
+			bg := root.Content().(*fyne.Container).Objects[0].(*background)
+			l.backgroundScreenMap[bg] = screen
+			if screen == l.screens.Primary() {
+				l.primaryWin = root
+				if l.bar == nil && l.widgets == nil && l.mouse == nil {
```

Comment:
No this root code refactors existing windows and widgets.  The code above creates windows and widgets.

Reply:
as noted above the 4 lines are literally identical

Reply:
Actually they aren't duplicates - the duplication is already factored out in createRoot() - this code is simply there to refactor a screen that already exists into a root screen rather than create one.

Reply:
Gotcha you weren't referring to the above code but the widget code